### ISSUE TITLE
Incorrect eta-reduction suggestion in the presence of RankNTypes

### ISSUE THREAD

**Issue by RyanGlScott** [ndmitchell/hlint#514]
`hlint` warns about the following code:

```haskell
{-# LANGUAGE RankNTypes #-}
module Foo where

type A = forall a. a -> a

f :: A -> ()
f x = undefined x
```
```
$ hlint Foo.hs 
Foo.hs:7:1: Warning: Eta reduce
Found:
  f x = undefined x
Perhaps:
  f = undefined

1 hint
```

This suggestion is incorrect, however, since using `f = undefined` will cause the code to no longer typecheck:

```
$ /opt/ghc/8.4.3/bin/ghc Foo.hs 
[1 of 1] Compiling Foo              ( Foo.hs, Foo.o )

Foo.hs:7:5: error:
    • Cannot instantiate unification variable ‘a0’
      with a type involving foralls: A -> ()
        GHC doesn't yet support impredicative polymorphism
    • In the expression: undefined
      In an equation for ‘f’: f = undefined
  |
7 | f = undefined
  |     ^^^^^^^^^
```
